File: backend/src/main/java/com/example/employeemanagement/service/DepartmentService.java
package com.example.employeemanagement.service;

import com.example.employeemanagement.exception.ResourceNotFoundException;
import com.example.employeemanagement.model.Department;
import com.example.employeemanagement.repository.DepartmentRepository;
import com.example.employeemanagement.repository.EmployeeRepository;
import java.util.List;
import java.util.Optional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

/** This class represents the service for departments. */
@Service
public class DepartmentService {

    @Autowired
    private DepartmentRepository departmentRepository;

    @Autowired
    private EmployeeRepository employeeRepository;

    /**
     * Get all departments with employees.
     */
    public List<Department> getAllDepartments() {
        return departmentRepository.findAllWithEmployees();
    }

    /**
     * Get department by ID or throw exception.
     */
    public Department getDepartmentOrThrow(Long id) {
        return departmentRepository.findByIdWithEmployees(id)
                .orElseThrow(() ->
                        new ResourceNotFoundException("Department not found with id: " + id));
    }

    /**
     * Create a new department.
     */
    public Department createDepartment(String name) {
        Department department = new Department();
        department.setName(name);
        return departmentRepository.save(department);
    }

    /**
     * Update existing department.
     */
    public Department updateDepartment(Long id, String name) {
        Department department = getDepartmentOrThrow(id);
        department.setName(name);
        return departmentRepository.save(department);
    }

    /**
     * Delete department with validation.
     */
    public void deleteDepartment(Long id) {
        Department department = getDepartmentOrThrow(id);

        long employeeCount = employeeRepository.countByDepartmentId(id);
        if (employeeCount > 0) {
            throw new IllegalStateException(
                    "Cannot delete department with " + employeeCount + " employees. " +
                            "Reassign or remove employees first."
            );
        }

        departmentRepository.delete(department);
    }

    public long countEmployeesInDepartment(Long id) {
        return employeeRepository.countByDepartmentId(id);
    }
}

File: frontend/src/services/departmentService.js
import axios from 'axios';

const API_URL = 'https://employee-management-app-gdm5.onrender.com/api/departments';

// Get all departments
export const getAllDepartments = async () => {
  const response = await axios.get(API_URL);
  return response.data;
};

// Get department by ID
export const getDepartmentById = async id => {
  const response = await axios.get(`${API_URL}/${id}`);
  return response.data;
};

// Add a new department
export const addDepartment = async department => {
  const response = await axios.post(API_URL, department);
  return response.data;
};

// Update an existing department
export const updateDepartment = async (id, department) => {
  const response = await axios.put(`${API_URL}/${id}`, department);
  return response.data;
};

// Delete a department
export const deleteDepartment = async id => {
  await axios.delete(`${API_URL}/${id}`);
};
